### PR TITLE
Add support of static content

### PR DESCRIPTION
**Pull request by bryce-anderson**
the server package `staticcontent` provides some tools for serving up
static content from either a resource or from a file.

* Configuration allows for caching (only one very basic example right now).
* Some support for Range headers.

### CODE REVIEW

File: core/src/main/scala/org/http4s/StaticFile.scala [hunk 1]
```diff
@@ -67,9 +67,7 @@ object StaticFile {
 
   def fromFile(f: File, start: Long, end: Long, buffsize: Int, req: Option[Request])
                         (implicit es: ExecutorService): Option[Response] = {
-
-    if (start < 0 || end < start || buffsize <= 0)
-      throw new Exception(s"start: $start, end: $end, buffsize: $buffsize")
+    require (start >= 0 && end > start && buffsize > 0, s"start: $start, end: $end, buffsize: $buffsize")
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/StaticFile.scala]: Exceptions make me sad.  Would it be sensible to return `Some(InternalServerError)` instead?

Comment (bryce-anderson) [inline core/src/main/scala/org/http4s/StaticFile.scala]: I don't like exceptions either. However, I think ```Some(InternalServerError)``` is not the solution for two main reasons:
* Unless they are explicitly logging 5xx's (which they should...) the failure will be silent and even then they won't get a descriptive message making it hard to figure out they are not checking the arguments to StaticFile. 
* If the route wants to attempt to recover it has to begin peeking at the status code which is never going to be intuitive or worth it.

This situation reminds me of places like https://github.com/scalaz/scalaz-stream/blob/292230aaa158bb4141da2a1aa15c23df818209df/src/main/scala/scalaz/stream/Process.scala#L845 (among others) where you should be verifying the arguments before calling the function.
Maybe there is another way to deal with this?

File: server/src/main/scala/org/http4s/server/staticcontent/ResourceService.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+                    bufferSize: Int = 256*1024,
+                    executor: ExecutorService = Strategy.DefaultExecutorService,
+                    cacheStartegy: CacheStrategy = NoopCacheStrategy)
+
+  /** Make a new [[org.http4s.server.HttpService]] that serves static files. */
+  private[staticcontent] def apply(config: Config): HttpService = Service.lift { req =>
+    val uri = req.uri
+    if (!uri.path.startsWith(config.pathPrefix)) Task.now(None)
+    else OptionT(
+          StaticFile.fromResource(sanitize(config.basePath + File.separator + getSubPath(uri, config.pathPrefix)))
```
Comment (d6y) [inline server/src/main/scala/org/http4s/server/staticcontent/ResourceService.scala]: Just checking: is `File.separator` the right value to use for a resource? My impression was that a resource is handled like a URI, and the separator is `/`.

Comment (bryce-anderson) [inline server/src/main/scala/org/http4s/server/staticcontent/ResourceService.scala]: I think you're right. Just happens that most everything but windows uses ```/```. I'll change it. 
Thanks for the review.

File: core/src/main/scala/org/http4s/StaticFile.scala (hunk 1 above)
Comment (bryce-anderson) [inline core/src/main/scala/org/http4s/StaticFile.scala]: One thing to consider is making the return types some form of ```Task```. That will put the error handling in line with the rest of the library. It makes me sad to make things asynchronous when they need not be, but I think thats justified considering the pervasiveness of ```Task``` in the library and sparsity of exceptions. I think that might be a separate PR if we go that route.

**Comment by bryce-anderson** [comment]
Where does this PR stand? Does it need more work? More discussion?

File: core/src/main/scala/org/http4s/StaticFile.scala (hunk 1 above)
Comment (rossabaker) [inline core/src/main/scala/org/http4s/StaticFile.scala]: So it would need to be a `Task[Option[Response]]` then?  With taskiness only to express failure?

**Comment by rossabaker** [comment]
I think this is progress.  I'm not moved by any of the options that address my complaint.  :+1:

File: core/src/main/scala/org/http4s/StaticFile.scala (hunk 1 above)
Comment (bryce-anderson) [inline core/src/main/scala/org/http4s/StaticFile.scala]: Yes, that is basically it. I'm not that excited about the idea either.
